File: src/python/algorithms/dynamicProgramming/uniqueBSTs2/uniquebsts2.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def generate_trees(n):
    if n == 0:
        return []
    def generate(left, right, memo):
        if left > right:
            return [None]
        if (left, right) in memo:
            return memo[(left, right)]
        result = []
        for root in range(left, right + 1):
            for left_tree in generate(left, root - 1, memo):
                for right_tree in generate(root + 1, right, memo):
                    node = TreeNode(root, left_tree, right_tree)
                    result.append(node)
        memo[(left, right)] = result
        return result
    return generate(1, n, {})

def preorder(root):
    if not root:
        return ['null']
    return [str(root.val)] + preorder(root.left) + preorder(root.right)

if __name__ == "__main__":
    test_cases = [0, 1, 2, 3]
    for i, n in enumerate(test_cases, 1):
        print(f"Test case {i}: n = {n}")
        trees = generate_trees(n)
        print("Total unique BSTs:", len(trees))
        for idx, root in enumerate(trees, 1):
            print(f"Tree {idx}: {' '.join(preorder(root))}")

File: src/javascript/algorithms/dynamicProgramming/uniqueBSTs2/uniqueBSTs2.js
class TreeNode {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function generateTrees(n) {
    if (n === 0) return [];
    const memo = {};
    function generate(left, right) {
        if (left > right) return [null];
        const key = `${left},${right}`;
        if (memo[key]) return memo[key];
        const result = [];
        for (let root = left; root <= right; root++) {
            for (const leftTree of generate(left, root - 1)) {
                for (const rightTree of generate(root + 1, right)) {
                    result.push(new TreeNode(root, leftTree, rightTree));
                }
            }
        }
        memo[key] = result;
        return result;
    }
    return generate(1, n);
}

function preorder(root) {
    if (!root) return ['null'];
    return [root.val.toString()].concat(preorder(root.left), preorder(root.right));
}

// Test cases
const testCases = [0, 1, 2, 3];
testCases.forEach((n, i) => {
    console.log(`Test case ${i + 1}: n = ${n}`);
    const trees = generateTrees(n);
    console.log('Total unique BSTs:', trees.length);
    trees.forEach((root, idx) => {
        console.log(`Tree ${idx + 1}: ${preorder(root).join(' ')}`);
    });
});
